velox/docs/develop/spilling.rst

```diff
@@ -0,0 +1,480 @@
+with a subset of rows from that partition to join. For the hash probe operator,
+it itself is not spillable but it needs to spill the input rows if the
+associated partition has been spilled by the hash build. We will discuss this
+further in the hash join spilling section.
+
+.. code-block:: c++
+
+    void Spiller::spill(uint32_t partition, const RowVectorPtr& spillVector);
+
+`SpillFileList <https://github.com/facebookincubator/velox/blob/main/velox/exec/Spill.h#L155>`_ and `SpillFile <https://github.com/facebookincubator/velox/blob/main/velox/exec/Spill.h#L58>`_
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+There is one SpillFileList object for each spill partition to manage its spill
+files. Each spill file is managed by one SpillFile object which provides the
+low level io operations with the storage system through Velox file system
+interface. On the spill path, the SpillFileList object takes a row vector as
+input, and internally it creates a VectorStreamGroup to serialize the row
+vector and write out the serialized byte stream into the currently opened spill
+file through the corresponding SpillFile object. The SpillFileList object will
+also switch to the next spill file if the current file grows beyond the target
+file size. On the restore path, the SpillFile object reads the serialized byte
+stream from the underlying storage system, and then uses VectorStreamGroup to
+deserialize the byte stream into row vectors.
+
+Spill Triggers
+--------------
+
+Spilling will be integrated with the Velox memory management system as a way to
+reclaim memory when the system lacks memory. Whenever an operator can’t
+allocate or reserve new memory, the memory arbitrator will select a number of
+Velox tasks to shrink their memory usage to satisfy the new memory allocation
+or reservation request. Each selected task will try to reclaim memory from its
+spillable operators. The latter in turn frees up memory by spilling out (part)
+of its memory state to disk. The integration of spilling with the memory
+management system is being under development.
+
+We also support to trigger spilling if the spillable operator's memory usage
+exceeds a configurable limit. This helps us to prevent query from running out
+of memory without the memory arbitration support. Note that the spilling itself
+can’t totally prevent out of memory as the last memory allocation that exceeds
+the memory limit, can be made from any operator in a query plan not always from
+the spillable one.
+
+Spill Parameters
+----------------
+Spill File Size
+^^^^^^^^^^^^^^^
+Given the same amount of spill data, the spill file size determines the number
+of spill files on disk. On the one hand, we should avoid generating too many
+small spill files as it might overload the metadata service of the storage
+system. On the other hand, we also want a sufficient number of spill files to
+parallelize the restore works. For example, to build a hash table from a
```

Comment:
Are you using "works" as a noun (in plural form)? Not sure if that's valid English.